### PR TITLE
Loki: Rewrite getHighlighterExpressionsFromQuery to use Loki parser

### LINKED ISSUE #53094: Loki: Rewrite getHighlighterExpressionsFromQuery to use Loki parser
**Level: medium**

In `public/app/plugins/datasource/loki/query_utils.ts` in `getHighlighterExpressionsFromQuery` we are using complicated logic and regex to figure out what strings should be highlighted. Instead of that, we could use newly introduce Loki parser, parse the string and check for all `LineFilters`. Then we can decide based on type of operator - exact match (`=`) or regex match (`~=`) if it needs  `escapeRegExp` or not. 

The similarly logic of using parser to get `LineFilter` is in `getLineFiltersPositions` in `public/app/plugins/datasource/loki/addToQuery.ts`

### PR DESCRIPTION
**Pull request by matyax**
**What this PR does / why we need it**:

This PR uses the parser from `lezer-logql` to replace a custom implementation using regular expressions.

**Which issue(s) this PR fixes**:

Fixes #53094 

**Special notes for your reviewer**:

Used the existing unit tests to make sure the original behavior was not lost in the refactor.

### CODE REVIEW

**Review by ivanahuckova — APPROVED** [review]
Beautiful readable code 🤩. LGMT!

File: public/app/plugins/datasource/loki/query_utils.ts
Comment (svennergr) [inline public/app/plugins/datasource/loki/query_utils.ts:44]: I noticed, that this will break the highlighting if we are filtering for a string containing quotes inside backticks:
image

I guess we can remove the regex matching here, determine the `filterTerm` by using `input.substring(string.from+1, string.to-1);` and `backtickedTerm` by just checking `input.substring(string.from, string.from+1);`?

Feels a bit hacky but should do the thing. Or what do you think @ivanahuckova and @matyax?

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/query_utils.ts:44]: Yeah, nice edge case catch. We should add test for this scenario. 

So the issue here is that in this case we match both `backtickedTerm` and  `quotedTerm` and because `term = quotedTerm || backtickedTerm`, it is going to match the `quotedTerm` inside. 

I am wondering if we should just determine if it is  `backtickedTerm` or `quotedTerm` based on `filterTerm[0]` character? It is always going to be just ` or ".

Comment (svennergr) [reply public/app/plugins/datasource/loki/query_utils.ts:44]: Yea, the regex matching there will always cause some kind of issues, because quoted terms can contain backticks, and backticked terms can contain quotes.

> I am wondering if we should just determine if it is `backtickedTerm` or `quotedTerm` based on `filterTerm[0]` character? It is always going to be just ` or ".

Yea, that would essentially be the same as I was trying to say with 
> `backtickedTerm` by just checking `input.substring(string.from, string.from+1);`?

Comment (matyax) [reply public/app/plugins/datasource/loki/query_utils.ts]: Good catch! If I read the code correctly, the previous implementation also suffered from this edgecase, since that part was unchanged?

Anyway, in the last commit https://github.com/grafana/grafana/pull/53316/commits/977ab6ee77b7e258a5a223e7451065481564f0b3 I propose an alternative approach, where I updated the regex with beginning and end delimiters, to catch the actual quote type used for the whole term. Let me know what you think, and thanks for the feedback!

Comment (svennergr) [reply public/app/plugins/datasource/loki/query_utils.ts]: Yep, definitely the previous implementation suffered from this as well.

Hm, https://github.com/grafana/grafana/commit/977ab6ee77b7e258a5a223e7451065481564f0b3 is fixing this, yea. I am unsure if I really prefer the RegEx matching over character matching just to determine, if it is a backticked filterterm. Performance-wise (even though it is not a super big issue) I would prefer checking based on `filterTerm[0]`, but we could also just stick to the solution now.

Comment (matyax) [reply public/app/plugins/datasource/loki/query_utils.ts]: We would also need to check `filterTerm[filterTerm.length - 1]` to be sure the term is properly quoted, right? Otherwise `"term" and `\`term` would incorrectly match.

I think the regex abstracts all those conditionals in a more readable way, opposite to the string-comparison approach.

Regarding performance, do you foresee any particular issue with it? Is this a function that gets called often?

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/query_utils.ts]: > We would also need to check filterTerm[filterTerm.length - 1] to be sure the term is properly quoted, right?

If it wasn't properly quoted, then parser shouldn't return `String` as string should be sterted-ended with the same set of quotes. But it does, so I am looking at the parser, trying to fix it these.

Comment (matyax) [reply public/app/plugins/datasource/loki/query_utils.ts]: Perfect, thanks!

Comment (matyax) [reply public/app/plugins/datasource/loki/query_utils.ts]: Updated.

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/query_utils.ts]: Btw here is the fix https://github.com/grafana/lezer-logql/pull/14. When it is merged, it will automatically add this new updated version `0.0.16` to `npm` and then we can update it in this `grafana/grafana` repo.

Comment (svennergr) [reply public/app/plugins/datasource/loki/query_utils.ts]: I think with the parser-change we don't need to check the last character and can just go with the first character as backtick-or-not determination, right?

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/query_utils.ts]: Version `0.0.16` is out now https://www.npmjs.com/package/@grafana/lezer-logql/v/0.0.16.

Comment (matyax) [reply public/app/plugins/datasource/loki/query_utils.ts]: By the way, I'm waiting for your re-approval after the recent changes to merge. Not sure why it did not dismiss the original approve with further commits.

**Review by svennergr — APPROVED** [review]
LGTM 👍

**Pull request merged**
